xen/arch/x86/slaunch.c

```diff
@@ -0,0 +1,279 @@
+#include <xen/types.h>
+#include <asm/e820.h>
+#include <asm/intel_txt.h>
+#include <asm/page.h>
+#include <asm/processor.h>
+#include <asm/slaunch.h>
+#include <asm/tpm.h>
+#include <xen/init.h>
+#include <xen/mm.h>
+#include <xen/multiboot.h>
+
+/* SLB is 64k, 64k-aligned */
+#define SKINIT_SLB_SIZE  0x10000
+#define SKINIT_SLB_ALIGN 0x10000
+
+bool __initdata slaunch_active;
+uint32_t __initdata slaunch_param;
+
+static void __maybe_unused compile_time_checks(void)
+{
+    BUILD_BUG_ON(sizeof(slaunch_active) != 1);
+}
+
+int __init map_l2(unsigned long paddr, unsigned long size)
+{
+    unsigned long aligned_paddr = paddr & ~((1ULL << L2_PAGETABLE_SHIFT) - 1);
+    unsigned long pages = ((paddr + size) - aligned_paddr);
+    pages = ROUNDUP(pages, 1ULL << L2_PAGETABLE_SHIFT) >> PAGE_SHIFT;
+
+    if ( (aligned_paddr + pages * PAGE_SIZE) <= PREBUILT_MAP_LIMIT )
+        return 0;
+
+    if ( aligned_paddr < PREBUILT_MAP_LIMIT )
+    {
+        pages -= (PREBUILT_MAP_LIMIT - aligned_paddr) >> PAGE_SHIFT;
+        aligned_paddr = PREBUILT_MAP_LIMIT;
+    }
+
+    return map_pages_to_xen((unsigned long)__va(aligned_paddr),
+                            maddr_to_mfn(aligned_paddr),
+                            pages, PAGE_HYPERVISOR);
+}
+
+static uint32_t get_slb_start(void)
+{
+    /* The runtime computation relies on size being a power of 2 and equal to
+     * alignment. Make sure these assumptions hold. */
+    BUILD_BUG_ON(SKINIT_SLB_SIZE != SKINIT_SLB_ALIGN);
+    BUILD_BUG_ON(SKINIT_SLB_SIZE == 0);
+    BUILD_BUG_ON((SKINIT_SLB_SIZE & (SKINIT_SLB_SIZE - 1)) != 0);
+
+    /* Rounding any address within SLB down to alignment gives SLB base and
+     * slaunch_param points to SLRT inside SLB on AMD. */
+    return slaunch_param & (SKINIT_SLB_SIZE - 1);
+}
+
+void __init map_skinit_mem_regions(void)
+{
+    map_l2(TPM_TIS_BASE, TPM_TIS_BASE + TPM_TIS_SIZE);
+
+    map_l2(get_slb_start(), get_slb_start() + SKINIT_SLB_SIZE);
```

Comment:
`map_l2()` takes size as a parameter, not the end address:

```suggestion
    map_l2(TPM_TIS_BASE, TPM_TIS_SIZE);

    map_l2(get_slb_start(), SKINIT_SLB_SIZE);
```

Reply:
https://github.com/TrenchBoot/xen/compare/6c44eb9bcc4f73b6b98e6a6cfd50c0994c85580c..0eac7a817733f1f1d3a7ab853e49ad47473d0396#diff-1d266d1dff9daf02e019100b9c620a099561c98811a05a0b7581363b3c00ac39R58

Reply:
You missed the SLB size.

Reply:
![](http://www.sherv.net/cm/emoticons/fighting/beating-head-against-the-wall.gif)
Apparently this is the most complicated part of this PR. https://github.com/TrenchBoot/xen/compare/0eac7a817733f1f1d3a7ab853e49ad47473d0396..0bac2f30b7bb5c9f4946a0ea27f5d9fc816e80a0